### ISSUE TITLE
pthread 3.0.4 causes ERR_CONNECTION_RESET on PHP 7RC3

### ISSUE THREAD

**Issue by Subdartha** [krakjoe/pthreads#490]
Hello,
I'm trying to get anything from pthread working on new PHP 7RC3:
```php
<?php
ini_set('display_errors',1);
ini_set('display_startup_errors',1);
error_reporting(-1);

class CMyThread extends Thread
{
  public function run() {}
}

$mythread = new CMyThread();
if ($mythread->start())
  $mythread->join();
echo "OK";
?>
```
If instead of main block I put simple `phpinfo()`, I see pthreads installed and activated:
```
PHP Version 7.0.0RC3
pthreads
Version	3.0.4
```
There are no errors loading the extension (I tried removing the file from there, it reports error then).
If I try to run the block which I showed, I get ERR_CONNECTION_RESET, and error in Windows applications log:
```
Faulting application name: httpd.exe, version: 2.4.16.0, time stamp: 0x55ae50e3
Faulting module name: php7apache2_4.dll, version: 7.0.0.0, time stamp: 0x55fb7afe
Exception code: 0xc0000005
Fault offset: 0x00002b81
Faulting process id: 0x8210
Faulting application start time: 0x01d0f4853e6ab62e
Faulting application path: C:\WAMP\apache2\bin\httpd.exe
Faulting module path: C:\WAMP\php\php7apache2_4.dll
Report Id: 7e97ea79-6078-11e5-bab3-80fa5b0f0470
```
Not sure which logs I could provide to help investigating this... or maybe there will be an idea what I'm doing wrong here. However, with 5.6 and pthreads v2 I didn't have such problem.

Thanks in advance!

**Comment by krakjoe** [comment]
I realise, this is not what you were looking for.

I'm sorry about that, but using threads in a web server is a really terrible idea. I'll explain this in detail in an upcoming blog post, when I've time to write it.

Suffice to say for now, that this has to happen, for my own sanity. If you want some insight into why I have taken this decision, [read this](http://blog.krakjoe.ninja/2014/10/but-is-it-web-scale.html).

**Comment by Subdartha** [comment]
Thank you for the feedback, Joe!

Somehow such restriction seems to me a little too strict, and here's why.
In that article you're using the argument, that even a relatively small number of clients would overload the hardware with too many threads running simultaneously. I completely agree with your warning that it should be avoided; however this isn't the only way what the server can be used for. For instance, my project is indended to be used at its top by small team of developers - not more than 30, that's the business case, and the idea of using multi-threading came to me, when I didn't want to establish 2 connections to the sockets server from the clients - and actually you have the sockets server in your examples ;) So max number of threads would be 60, which the hardware can definitely chew.

Bottom line is, in some cases the use of multi-threading, IMHO, could be appropriate even on web-servers, given the author of the application understands the consequences and restrictions. I.e. I believe I would benefit from it in my particular case, and with such commit it's becoming less possible.

In any case, - thanks for your hard work and attention to reported bugs!

**Comment by krakjoe** [comment]
There are many many other arguments against running threads inside a web server, there are no decent ones in support of it.

It's not just a case of what is a good idea, it's also a case of what you can make stable and robust. Let's take apache for an example, although the same applies to fpm; In apache you have multiple processes with multiple threads, processes are controlled via signals. Zend wants to block signals while some criticial operations occur (restart/reload and all the network related signals the processes might recv), depending on the build options of php, sometimes they are blocked, sometimes they get through. The rules on handling signals in a process that is both multi-process and multi-threaded are complicated, and you cannot satisfy all their requirements from inside of PHP.

In CLI, we are as safe as can be, we cannot make the same assumptions *anywhere* else.

I allowed execution anywhere at first, to lower the barrier to entry. I was wrong to do that, the barrier to entry is high no matter what I do.

I want you to be able to write robust, reliable, predictable code with pthreads v3, this simply can't happen in the context of a web server, it doesn't make sense to try.

**Comment by Subdartha** [comment]
Thanks for thorough argumentation!
